### ISSUE TITLE
Install error alongside Guzzle 7

### ISSUE THREAD

**Issue by diablo341** [Nexmo/nexmo-php#231]
I read an older post about HTTPlug/Guzzle 6 #5  , but I'm not sure what to do with my error.  I run composer require nexmo/client and get the following error:

> λ composer require nexmo/client                                                   
> Using version ^2.0 for nexmo/client                                               
> ./composer.json has been updated                                                  
> Loading composer repositories with package information                            
> Updating dependencies (including require-dev)                                     
> Your requirements could not be resolved to an installable set of packages.        
>                                                                                   
>   Problem 1                                                                       
>     - Installation request for nexmo/client ^2.0 -> satisfiable by nexmo/client[2.
> 0.0, v2.2.0-RC1, v2.2.0-RC2, v2.2.0.x-dev].                                       
>     - Can only install one of: guzzlehttp/guzzle[6.5.x-dev, 7.0.1].               
>     - Can only install one of: guzzlehttp/guzzle[7.0.1, 6.5.x-dev].               
>     - Can only install one of: guzzlehttp/guzzle[6.5.x-dev, 7.0.1].               
>     - Conclusion: install guzzlehttp/guzzle 6.5.x-dev                             
>     - Installation request for guzzlehttp/guzzle (locked at 7.0.1, required as ^7.
> 0) -> satisfiable by guzzlehttp/guzzle[7.0.1].                                    
>                   
> Installation failed, reverting ./composer.json to its original content.

I have guzzlehttp/guzzle 7.0 already installed for another package.  Anything I can do to fix the install?

I also tried composer require nexmo/client-core "^2.0" and get the following error:

> λ composer require nexmo/client-core "^2.0"                                       
> ./composer.json has been updated                                                  
> Loading composer repositories with package information                            
> Updating dependencies (including require-dev)                                     
> Your requirements could not be resolved to an installable set of packages.        
>                                                                                   
>   Problem 1                                                                       
>     - nexmo/client-core v2.2.0.x-dev requires php-http/client-implementation ^1.0 
> -> no matching package found.                                                     
>     - nexmo/client-core v2.2.0-RC2 requires php-http/client-implementation ^1.0 ->
>  no matching package found.                                                       
>     - nexmo/client-core v2.2.0-RC1 requires php-http/client-implementation ^1.0 ->
>  no matching package found.                                                       
>     - nexmo/client-core 2.1.0 requires php-http/client-implementation ^1.0 -> no m
> atching package found.                                                            
>     - nexmo/client-core 2.0.0 requires php-http/client-implementation ^1.0 -> no m
> atching package found.                                                            
>     - Installation request for nexmo/client-core ^2.0 -> satisfiable by nexmo/clie
> nt-core[2.0.0, 2.1.0, v2.2.0-RC1, v2.2.0-RC2, v2.2.0.x-dev].                      
>                                                                                   
> Potential causes:                                                                 
>  - A typo in the package name                                                     
>  - The package is not available in a stable-enough version according to your minim
> um-stability setting                                                              
>    see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more detai
> ls.                                                                               
>  - It's a private package and you forgot to add a custom repository to find it    
>                                                                   
> Installation failed, reverting ./composer.json to its original content.

Shouldn't guzzle 7 include the php-http/client-implementation package?

@dragonmantank any ideas?

I've attached my composer.json file --> [composer.json.pdf](https://github.com/Nexmo/nexmo-php/files/4942375/composer.json.pdf)

Thanks,
John

**Comment by dragonmantank** [comment]
I haven’t tested the composer.json file, but since ‘Nexmo/client’ has a hard dependency on the guzzle 6 adapter, and your composer.json is requesting Guzzle 7, it’s hitting the same problem as #5. 

At this time, I think the only option is to drop back to Guzzle 6. There is no Guzzle 7 adapter available from HTTPlug, which means it’s not compatible with our package :( 

I’ll see if I can find out what the status of a Guzzle 7 adapter is in the meantime, and will leave this ticket open.

**Comment by dragonmantank** [comment]
@diablo341 Good news! Guzzle 7 is PSR-18 compatible, and it looks like we have a decent upgrade path to just move to directly supporting PSR-18 HTTP clients directly. I _believe_ I can do this as a non-breaking change too.

I don't recommend running testing code in production, but if you'd like to help test this fix out, I've pushed it to the upcoming v2.2.0 branch. Instead of just requiring `nexmo/client`, can you instead use:

```json
{
    // ...
    "minimum-stability": "dev",
    "require": {
        // ...
        "nexmo/client-core": "v2.2.0.x-dev",
    }
}
```
You can then use the same fix as #5, and just create a Guzzle 7 client and pass it in as the 3rd parameter into our constructor:

```php
$guzzle = new \GuzzleHttp\Client();
$nexmo = new Client(
    new Basic(getenv('NEXMO_API_KEY'), getenv('NEXMO_API_SECRET')),
    [],
    $guzzle
);
```

Everything else should work as before. If I can't confirm that this is backward-compatible, this change _will_ be in v3.0.0, which is coming out relatively soon after v2.2.0 is released.

**Comment by diablo341** [comment]
I will test on my end...

Thanks!
John

**Comment by diablo341** [comment]
Testing worked fine with guzzle 7.0.1 and v2.2.0-dev.

Thanks!

**Comment by dragonmantank** [comment]
This should now be resolved with the release of 2.2.0 proper. It now can be handled like [any other Guzzle conflicting package](https://github.com/Nexmo/nexmo-php#composer-installation-fails-due-to-guzzle-adapter).
